### PR TITLE
#254 - [macOS] Possible NPE in Combo.mouseDown

### PR DESCRIPTION
**Pull request by tmssngr**
better cache the used display before it gets unset

### CODE REVIEW

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Combo.java [hunk 1]
```diff
@@ -1146,15 +1146,16 @@ void menuDidClose(long id, long sel, long menu) {
...
-	// popup window. Fix is to retain the view before letting Cocoa track
-	// the mouse events.
+	// popup window. Fix is to retain the display and view before letting Cocoa
+	// track the mouse events.
+
+	final Display display = this.display;
+	final NSView view = this.view;
 
 	display.sendPreExternalEventDispatchEvent();
-	// 'view' will be cleared if disposed during the mouseDown so cache it.
```
Comment (lshanmug) [inline bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Combo.java:1153]: Let's retain the comment and variable name as they don't need any changes.

**Review by lshanmug — CHANGES_REQUESTED** [review]

**Comment by SyntevoAlex** [comment]
I'd like to see a unit test here.
Most likely, if you dispose `Combo` in its `SWT.MouseDown` listener, it will reproduce reliably.

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Combo.java (hunk 1 above)
Comment (SyntevoAlex) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Combo.java:1153]: I would say that this is a matter of taste. And these a better left to author, if there's no obvious harm.

**Comment by tmssngr** [comment]
I've tried to add a unit test, but failed, because I was not able to send mouse events correctly, so this part of the code would be executed. See https://github.com/eclipse-platform/eclipse.platform.swt/compare/master...syntevo:eclipse.platform.swt:feature/254-macos-npe-mousedown-failed-test. I've also tried with `display.post`, but then other unit tests broke.

**Comment by SyntevoAlex** [comment]
If creating an automated test is difficult, then a manual snippet is just fine.

**Comment by tmssngr** [comment]
Is it OK now?

**Comment by SyntevoAlex** [comment]
The code comments were misplaced in your patch.
I fixed that and added commits to your branch. Please review and squash.

**Comment by SyntevoAlex** [comment]
The code comments were misplaced in your patch.

The old top-level comment was in fact explaining why `view.retain()` is needed.
You edited it incorrectly, `display` has nothing to do with it.

The `so cache it` code comment got misplaced after you moved variable declaration.

I fixed that and added commits to your branch. Please review and squash.

**Comment by SyntevoAlex** [comment]
Regarding the failing `test_setFocus_withVisibleAndInvisibleChild` test, see https://github.com/eclipse-platform/eclipse.platform.swt/issues/616#issuecomment-1616669535

File: bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Combo.java (hunk 1 above)
Comment (SyntevoAlex) [reply bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Combo.java:1153]: Not sure if reply is expected here. I'll resolve the conversation to merge PR.

**Comment by SyntevoAlex** [comment]
In master now.

**Comment by elsazac** [comment]
Verified on :

Eclipse SDK
Version: 2023-09 (4.29)
Build id: I20230703-1800
OS: Mac OS X, v.13.4.1, aarch64 / cocoa
Java vendor: Eclipse Adoptium
Java runtime version: 20+36
Java version: 20

**Pull request merged**